File: trie/trie_using_array.js
class Node {
	constructor() {
		this.array = new Array(26).fill(null)
		this.eow = false
	}
}

class Trie {
	constructor() {
		this.root = new Node()
	}

	insert(word) {

		let node = this.root

		for (let i=0; i<word.length; i++) {
			
			const char_index = word[i].charCodeAt() - 'a'.charCodeAt()

			if (node.array[char_index] === null) {				
				node.array[char_index] = new Node()				
			}

			if (i === (word.length-1)) {
				 node.array[char_index].eow = true
			}

			node = node.array[char_index]
		}
	}

	search(word) {
		let node = this.root

		for (let i=0; i<word.length; i++) {
			
			const char_index = word[i].charCodeAt() - 'a'.charCodeAt()

			if (node.array[char_index] === null) {
				return false
			}

			if ( i === (word.length-1) && node.array[char_index].eow === true) {				
				return true			
			}
			node = node.array[char_index]
		}

		return false
	}
}

/*
   {  array : [   ], eow: false}
*/

const trie = new Trie()
trie.insert('ab')

console.log(trie.search('ba'))

File: trie/trie_using_array.py
class Node:
	def __init__(self):
		self.children = [None] * 26
		self.end_of_word = False

class Trie:
	def __init__(self):
		self.root = Node()

	def insert(self, word):
		
		node = self.root

		for i in range(len(word)):				
			char_ascii_code = ord(word[i])
			char_trie_index = char_ascii_code - ord('a')
	
			if node.children[char_trie_index] is None:
				node.children[char_trie_index] = Node()

			if i == len(word)-1:
				node.children[char_trie_index].end_of_word = True

			node = node.children[char_trie_index]

	def search(self, word):
		node = self.root

		for i in range(len(word)):
			char_ascii_code = ord(word[i])
			char_trie_index = char_ascii_code - ord('a')
			
			if node.children[char_trie_index] is None:
				return False

			if i == len(word)-1 and node.children[char_trie_index].end_of_word == True:
				return True

			node = node.children[char_trie_index]

		return False

	def delete(self, word):
		node = self.root

		for i in range(len(word)):
			char_ascii_code = ord(word[i])
			char_trie_index = char_ascii_code - ord('a')

			if node.children[char_trie_index] is None:
				return False
			
			if i == len(word)-1 and node.children[char_trie_index].end_of_word == True:
				node.children[char_trie_index].end_of_word = False
				return True

			node = node.children[char_trie_index]

		return False

trie = Trie()
# trie.insert('aa')
# trie.insert('aant')
trie.insert('apple')
# trie.insert('apply')

print(trie.root.children[0][0])
# print(trie.search('apple'))
# print(trie.delete('aa'))
# print(trie.search('aa'))